File: src/java/test/src/androidTest/java/com/tencent/wcdbtest/winq/StatementCreateIndexTest.java
// Created by chenqiuwen on 2023/4/10.
//

/*
 * Tencent is pleased to support the open source community by making
 * WCDB available.
 *
 * Copyright (C) 2017 THL A29 Limited, a Tencent company.
 * All rights reserved.
 *
 * Licensed under the BSD 3-Clause License (the "License"); you may not use
 * this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 *       https://opensource.org/licenses/BSD-3-Clause
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.tencent.wcdbtest.winq;

import androidx.test.ext.junit.runners.AndroidJUnit4;
import com.tencent.wcdb.winq.*;
import static com.tencent.wcdbtest.base.WinqTool.winqEqual;

import org.junit.Test;
import org.junit.runner.RunWith;

@RunWith(AndroidJUnit4.class)
public class StatementCreateIndexTest {
    @Test
    public void test() {
        IndexedColumn index1 = new IndexedColumn("column1");
        IndexedColumn index2 = new IndexedColumn("column2").order(Order.Asc);
        String indexName = "index1";
        String tableName = "table1";

        winqEqual(new StatementCreateIndex().createIndex(indexName).on(tableName)
                        .indexedBy(new IndexedColumn[]{index1, index2}),
                "CREATE INDEX index1 ON table1(column1, column2 ASC)");

        winqEqual(new StatementCreateIndex().createIndex(indexName).of("testSchema").on(tableName)
                        .indexedBy(new IndexedColumn[]{index1, index2}),
                "CREATE INDEX testSchema.index1 ON table1(column1, column2 ASC)");

        winqEqual(new StatementCreateIndex().createIndex(indexName).unique().on(tableName).indexedBy(index2),
                "CREATE UNIQUE INDEX index1 ON table1(column2 ASC)");

        winqEqual(new StatementCreateIndex().createIndex(indexName).ifNotExist().on(tableName).indexedBy("newColumn"),
                "CREATE INDEX IF NOT EXISTS index1 ON table1(newColumn)");

        winqEqual(new StatementCreateIndex().createIndex(indexName).on(tableName)
                .indexedBy(new String[]{"column1", "column2"}).where(new Column("column1").ge(1)),
                "CREATE INDEX index1 ON table1(column1, column2) WHERE column1 >= 1");
    }
}

File: src/rust/examples/tests/winq/statement_create_index_test.rs
#[cfg(test)]
pub mod statement_create_index_test {
    use crate::base::winq_tool::WinqTool;
    use wcdb::winq::column::Column;
    use wcdb::winq::expression_operable::ExpressionOperableTrait;
    use wcdb::winq::indexed_column::IndexedColumn;
    use wcdb::winq::ordering_term::Order;
    use wcdb::winq::statement_create_index::StatementCreateIndex;

    #[test]
    pub fn test() {
        let index1 = IndexedColumn::new("column1");
        let index2 = IndexedColumn::new("column2");
        index2.order(Order::Asc);
        let index_name = "index1";
        let table_name = "table1";

        let mut indexed_column_vec: Vec<&IndexedColumn> = Vec::new();
        indexed_column_vec.push(&index1);
        indexed_column_vec.push(&index2);
        WinqTool::winq_equal(
            StatementCreateIndex::new()
                .create_index(index_name)
                .on(table_name)
                .indexed_by(indexed_column_vec),
            "CREATE INDEX index1 ON table1(column1, column2 ASC)",
        );

        let mut indexed_column_vec: Vec<&IndexedColumn> = Vec::new();
        indexed_column_vec.push(&index1);
        indexed_column_vec.push(&index2);
        WinqTool::winq_equal(
            StatementCreateIndex::new()
                .create_index(index_name)
                .of("testSchema")
                .on(table_name)
                .indexed_by(indexed_column_vec),
            "CREATE INDEX testSchema.index1 ON table1(column1, column2 ASC)",
        );

        let mut indexed_column_vec: Vec<&IndexedColumn> = Vec::new();
        indexed_column_vec.push(&index2);
        WinqTool::winq_equal(
            StatementCreateIndex::new()
                .create_index(index_name)
                .unique()
                .on(table_name)
                .indexed_by(indexed_column_vec),
            "CREATE UNIQUE INDEX index1 ON table1(column2 ASC)",
        );

        let mut column_names: Vec<String> = Vec::new();
        column_names.push("newColumn".parse().unwrap());
        WinqTool::winq_equal(
            StatementCreateIndex::new()
                .create_index(index_name)
                .if_not_exist()
                .on(table_name)
                .indexed_by_column_names(&column_names),
            "CREATE INDEX IF NOT EXISTS index1 ON table1(newColumn)",
        );

        let mut column_names: Vec<String> = Vec::new();
        column_names.push("column1".parse().unwrap());
        column_names.push("column2".parse().unwrap());
        let expression = Column::new("column1", None).ge(1);
        WinqTool::winq_equal(
            StatementCreateIndex::new()
                .create_index(index_name)
                .on(table_name)
                .indexed_by_column_names(&column_names)
                .r#where(expression),
            "CREATE INDEX index1 ON table1(column1, column2) WHERE column1 >= 1",
        );
    }
}
